# Credit Risk Evaluator

In [11]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression

from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix

## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [20]:
# Import the data

lend_df = pd.read_csv('./Resources/lending_data.csv')

lend_df.head (30)

loan_size  interest_rate  borrower_income  debt_to_income  \
0     10700.0          7.672            52800        0.431818   
1      8400.0          6.692            43600        0.311927   
2      9000.0          6.963            46100        0.349241   
3     10700.0          7.664            52700        0.430740   
4     10800.0          7.698            53000        0.433962   
5     10100.0          7.438            50600        0.407115   
6     10300.0          7.490            51100        0.412916   
7      8800.0          6.857            45100        0.334812   
8      9300.0          7.096            47400        0.367089   
9      9700.0          7.248            48800        0.385246   
10     9300.0          7.085            47300        0.365751   
11     9200.0          7.015            46600        0.356223   
12     9200.0          7.043            46900        0.360341   
13     9400.0          7.101            47400        0.367089   
14     8400.0          6.703            43700        0.313501   
15     9400.0          7.136            47700        0.371069   
16     8900.0          6.897            45500        0.340659   
17     8300.0          6.648            43200        0.305556   
18     8000.0          6.539            42100        0.287411   
19    10200.0          7.460            50800        0.409449   
20     5200.0          5.326            30700        0.022801   
21     9300.0          7.068            47100        0.363057   
22     7300.0          6.225            39200        0.234694   
23     9000.0          6.943            45900        0.346405   
24     8700.0          6.813            44700        0.328859   
25     8800.0          6.850            45100        0.334812   
26     8400.0          6.700            43700        0.313501   
27     7900.0          6.482            41600        0.278846   
28     9700.0          7.267            49000        0.387755   
29     8400.0          6.700            43600        0.311927   

    num_of_accounts  derogatory_marks  total_debt  loan_status  
0                 5                 1       22800            0  
1                 3                 0       13600            0  
2                 3                 0       16100            0  
3                 5                 1       22700            0  
4                 5                 1       23000            0  
5                 4                 1       20600            0  
6                 4                 1       21100            0  
7                 3                 0       15100            0  
8                 3                 0       17400            0  
9                 4                 0       18800            0  
10                3                 0       17300            0  
11                3                 0       16600            0  
12                3                 0       16900            0  
13                3                 0       17400            0  
14                3                 0       13700            0  
15                3                 0       17700            0  
16                3                 0       15500            0  
17                2                 0       13200            0  
18                2                 0       12100            0  
19                4                 1       20800            0  
20                0                 0         700            0  
21                3                 0       17100            0  
22                2                 0        9200            0  
23                3                 0       15900            0  
24                3                 0       14700            0  
25                3                 0       15100            0  
26                3                 0       13700            0  
27                2                 0       11600            0  
28                4                 0       19000            0  
29                3                 0 

In [17]:
df.info ()
#df.describe ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Random Forest Classifier will probably be more accurate due to it's knack for generally being a more precise model. Also, Logistical regression is known to have poor accuracy when predicting categorical outcomes from the features of the model, whereas RFC is known for it's success in this and is frequently used to calculate financial models quite commonly, this is most likely due to the fact it does not assume that data has a linear relationship and it will weigh certain features of the model and deem them more important than others. (Example: It might find that the debt to income ratio is more important than borrower income and accept a lower debt to income ratio that has a lower household income.) 

# Split the Data into Training and Testing Sets

Create the features DataFrame, X, by removing the loan_status column. Create y, the labels set, by using the loan_status column. Split the data into training and testing datasets by using the train_test_split function.

In [13]:
# Split the data into X_train, X_test, y_train, y_test

x=lend_df.drop("loan_status",axis =1)
y=lend_df["loan_status"].values

x_train,x_test,y_train,y_test =train_test_split(x,y,random_state =21)

## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

In [22]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression()
classifier.fit(x_train, y_train)

print(f"LR Training Data: {classifier.score(x_train, y_train)}")
print(f"LR Testing Data: {classifier.score(x_test, y_test)}")

LR Training Data: 0.992450818544504
LR Testing Data: 0.9908687577383409


In [29]:
# Train a Random Forest Classifier model and print the model score

scaler = StandardScaler().fit(x_train)
x_train_scaled = scaler.transform(x_train)
x_test_scaled = scaler.transform(x_test)

clf =RandomForestClassifier( random_state = 21).fit(x_train_scaled, y_train)

y_pred=clf.predict(x_test_scaled)

print(f'RFC Training Data : {clf.score(x_train_scaled,y_train)}')
print(f'RFC Testing Data: {clf.score(x_test_scaled,y_test)}')

RFC Training Data : 0.9975409272252029
RFC Testing Data: 0.9909203466776723


The RFC perfoms better overall, but with less difference than I was expecting. It's training score is much higher, with a score of 0.9975 compared to the LR score of 0.9925. However, with the testing data there is very little difference between the two scores. RFC sits at 0.99092, whilst LR sits at 0.99082, when you round to 4dp, there is neglible difference bewteen the two models in correctly predicting the outcome of the data. Perhaps this is due to the linear nature of the data having similar outcomes in each scenario. Eg, the size of the loan compared to income, or the debt compared to debt to income ratio, naturally LR would be good at prediciting these outcomes despite the fact they are categorical. Potentially having a categorical outcome with only two possibilites with a largely linear dataset is an efficient way to train datasets such as this. 